### PR TITLE
Warn if override specifier list order differs from inheritance order

### LINKED ISSUE #8373: Warn if order in override specifier is different than in inheritance list
Non-breaking part of https://github.com/ethereum/solidity/issues/8354

### PR DESCRIPTION
**Pull request by Marenz**
fixes #8373

### CODE REVIEW

File: libsolidity/analysis/OverrideChecker.cpp [hunk 1]
```diff
@@ -767,6 +767,39 @@ void OverrideChecker::checkOverrideList(OverrideProxy _item, OverrideProxyBySign
 			}
 		}
 	}
+	else if (specifiedContracts.size() > 1)
+	{
+		vector<int64_t> overrideSpecifierContractIds{};
```
Comment (erak) [inline libsolidity/analysis/OverrideChecker.cpp]:
```suggestion
		vector<int64_t> overrideSpecifierContractIds;
```
Is this recommended by the unified initialization "guide"?

**Comment by erak** [comment]
Looks good already!

**Review by ekpyron — CHANGES_REQUESTED** [review]
If we warn about this, then I think we should also adjust the order in the suggestion "Function needs to specify overridden contracts [contracts].". Currently that list will ultimately just be alphabetical.

**Comment by Marenz** [comment]
that's a good point @ekpyron 

@chriseth suggested to use `linearizedBaseContracts` instead of `baseContracts()`, however, the following test case creates some issues with that:

for this example, the reversed linearized contracts mismatch
```
    contract I {
      function f() external virtual {}
    }
    contract A is I {
      function f() external virtual override {}
    }
    contract B is I {}
    contract C is A, B {
      function f() external override(A, I) {}
    }
```

the reversed list (filtered by the contracts specified in override) is `I, A`

**Comment by ekpyron** [comment]
You should check in what order ``super.f()`` calls the bases in this case - that's probably the order given by ``linearizedBaseContracts`` and that's probably the order we should require the mentions to have. I.e. ``linearizedBaseContracts`` is the "correct" order...

**Comment by ekpyron** [comment]
(the question how ``super`` behaves is the core of https://github.com/ethereum/solidity/issues/8354 which - if I understand correctly - led to the decision to warn about this, so I'd say the required order that's not warned about should be precisely the order of ``super`` calls, right?)

**Comment by Marenz** [comment]
Alright, so it was just my intuition that thought this is wrong then.

**Comment by ekpyron** [comment]
> Alright, so it was just my intuition that thought this is wrong then.

Better actually check what ``super`` does - I always found ``super`` to be rather weird myself, so I'd need to think about what it really does in this particular case myself :-).

**Comment by ekpyron** [comment]
FYI I'm myself not actually sure about requiring any particular order for this at all :-). IIUC we still can't guarantee that ``super.f()`` will actually call the function in the first mentioned contract, if you inherit from the whole thing again in a weird way...

**Comment by Marenz** [comment]
Well, one motivation for this was that people thought the order in `override(a,b)` implies the order of inheritance. If we just require the same as  specified at least that particular misunderstanding is eliminated

**Comment by Marenz** [comment]
updated to use `linearizedBaseContracts`

File: libsolidity/analysis/OverrideChecker.cpp [hunk 2]
```diff
@@ -791,18 +822,28 @@ void OverrideChecker::checkOverrideList(OverrideProxy _item, OverrideProxyBySign
...
 
 	if (!missingContracts.empty())
+	{
+		vector<ContractDefinition const*> sortedMissingContracts;
+		for (size_t i = _item.contract().annotation().linearizedBaseContracts.size() - 1; i > 0; i--)
+		{
+			auto contract = _item.contract().annotation().linearizedBaseContracts[i];
+			if (missingContracts.count(contract))
+				sortedMissingContracts.emplace_back(contract);
+		}
```
Comment (ekpyron) [inline libsolidity/analysis/OverrideChecker.cpp:832]:
```suggestion
for (
	auto const* contract: _item.contract().annotation().linearizedBaseContracts |
	boost::adaptor::filter([](auto const* contract) -> bool { return missingContracts.count(contract); }
)
  sortedMissingContracts.emplace_back(contract);
```
?

File: libsolidity/analysis/OverrideChecker.cpp [hunk 3]
```diff
@@ -767,6 +766,38 @@ void OverrideChecker::checkOverrideList(OverrideProxy _item, OverrideProxyBySign
 			}
 		}
 	}
+	else if (specifiedContracts.size() > 1)
```
Comment (ekpyron) [inline libsolidity/analysis/OverrideChecker.cpp:769]: Wouldn't this be much easier, if you did it *after* you checked that the specified contracts match the expected contracts (and only if they match)?

Then you could just do something like this, couldn't you?
```
auto overrideIterator = _item.overrides()->overrides().begin();
for (auto const* expected: _item.contract().annotation.linearizedBaseContracts | boost::adaptor::filter([](auto const* base) -> bool { return expectedContracts.count(base); })
  if ((overrideIterator++)->annotation().referencedDeclaration != expected)
  {
     // warn
     break;
  }
```

**Review by ekpyron — COMMENTED** [review]
Some suggestions.

File: libsolidity/analysis/OverrideChecker.cpp (hunk 2 above)
Comment (Marenz) [reply libsolidity/analysis/OverrideChecker.cpp:832]: Your suggestion is no longer reversing the linearized list?

**Comment by Marenz** [comment]
Closing this for now as an in-office discussion suggests we're unsure about this feature

**Pull request closed without merging**